### DISCUSSION TITLE
io.lettuce.core.RedisURI.Builder#withPassword(java.lang.String) deprecation

### DISCUSSION THREAD

**Discussion by perlun** [redis/lettuce#1695] [Q&A] [answered]
Hi,

We're currently upgrading our application from Lettuce 5.3.x to 6.0.3, and one thing I noted (since we have "warnings as errors" enabled for our project) is that the `io.lettuce.core.RedisURI.Builder#withPassword(java.lang.String)` method is now marked as deprecated:

```java
        /**
         * Configures authentication.
         *
         * @param password the password
         * @return the builder
         * @deprecated since 6.0. Use {@link #withPassword(CharSequence)} or {@link #withPassword(char[])} avoid String caching.
         */
        @Deprecated
        public Builder withPassword(String password) {
```

I'm generally a quite curious person and hence, I wonder a bit about this. Given that `String` implements the `CharSequence` interface, I can trivially work around this by just casing my input `String` to `CharSequence` instead and call the `withPassword(CharSequence)` overload isntead.

However, this must clearly not be how Lettuce intends for me to write my code, so hence the question. Is the idea to force/suggest that my calling code shouldn't cache the password in memory or what's the thinking here? If so, what is the suggested way to retrieve the password in a safe way from "whereever it is kept"?

/cc @mp911de

**Comment by mp911de** [comment, marked as answer]
The reason is that `String` has a strong caching affinity and the JVM cannot be easily GC `String` instances. Therefore we suggest either using `char[]` or a custom `CharSequence` (`StringBuilder`, netty's `AsciiString`). Deprecating this method should raise awareness and indicate that there are better options for password handling than `String`.

**Reply by perlun** [reply]
Thanks @mp911de. :+1: I'll see what will be an appropriate action on our part.

(I guess it could make sense to copy what you just wrote to an FAQ of some sort or similar.)

**Reply by mp911de** [reply]
Happy to review a pull request.

**Reply by perlun** [reply]
Here we go: https://github.com/lettuce-io/lettuce-core/pull/1707

**Reply by haicoder** [reply]
Dose one password string will effect GC ? 
I don't understand your explanation, I'd appreciate it if you can explain it with more detail.

**Reply by perlun** [reply]
@haicoder Sorry for the late reply. The question is not about whether it _affects_ GC or not, but the problem is that the password will remain in memory since `String` instances are more likely to be cached and less likely to be eligible for garbage collection. Given that the password is sensitive information, it gives more room to a potential attacker to read the memory from the process in question and extract the password that way. _That_, from how I understand it, is the reason why Lettuce is deprecating this method.
